cpp/src/binary/jit/core/binop.cpp

```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright 2018-2019 BlazingDB, Inc.
+ *     Copyright 2018 Christian Noboa Mardini <christian@blazingdb.com>
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "binary/jit/core/launcher.h"
+#include "binary/jit/util/operator.h"
+#include "cudf.h"
+
+namespace gdf {
+namespace binops {
+namespace jit {
+
+    struct Option {
+        Option(bool state, gdf_error value)
+         : is_correct{state}, gdf_error_value{value}
+        { }
+
+        operator bool() {
+            return is_correct;
+        }
+
+        gdf_error get_gdf_error() {
+            return gdf_error_value;
+        }
+
+    private:
+        bool is_correct;
+        gdf_error gdf_error_value;
+    };
+
+    Option verify_scalar(gdf_scalar* scalar) {
+        if (scalar == nullptr) {
+            return Option(false, GDF_DATASET_EMPTY);
+        }
+        if (scalar->data == nullptr) {
+            return Option(false, GDF_DATASET_EMPTY);
+        }
+        if ((scalar->dtype <= GDF_invalid) || (N_GDF_TYPES <= scalar->dtype)) {
+            return Option(false, GDF_UNSUPPORTED_DTYPE);
+        }
+        return Option(true, GDF_SUCCESS);
+    }
+
+    Option verify_column(gdf_column* vector) {
+        if (vector == nullptr) {
+            return Option(false, GDF_DATASET_EMPTY);
+        }
+        if (vector->size == 0) {
+            return Option(false, GDF_SUCCESS);
+        }
+        if (vector->data == nullptr) {
+            return Option(false, GDF_DATASET_EMPTY);
+        }
+        if ((vector->dtype <= GDF_invalid) || (N_GDF_TYPES <= vector->dtype)) {
+            return Option(false, GDF_UNSUPPORTED_DTYPE);
+        }
+        return Option(true, GDF_SUCCESS);
+    }
+
+    Option verify_column(gdf_column* out, gdf_column* vax) {
+        auto result = verify_column(out);
+        if (!result) {
+            return result;
+        }
+        result = verify_column(vax);
+        if (!result) {
+            return result;
+        }
+        if (out->size < vax->size) {
+            return Option(false, GDF_COLUMN_SIZE_MISMATCH);
+        }
+        return Option(true, GDF_SUCCESS);
+    }
+
+    Option verify_column(gdf_column* out, gdf_column* vax, gdf_column* vay) {
+        auto result = verify_column(out);
```

Comment:
All of this custom error checking code is non-standard and should instead be replaced with `GDF_REQUIRE`.